### PR TITLE
CA-843 implement single BQ query for all expressions, relationships and all

### PR DESCRIPTION
**Pull request by dvoet**
https://broadworkbench.atlassian.net/browse/CA-843

implemented conversion of parse expressions to BigQuery sql statement, execute, convert results to what is expected from submission machinery

- [x] **Submitter**: Include the JIRA issue number in the PR description
- [x] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [x] **...and Orchestration's Swagger too!**
- [x] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * To Demo flag is set
  * Release Summary is filled out, if applicable
  * Add notes on how to QA
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [ ] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via Slack and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

**Review by davidangb — COMMENTED** [review]
I didn't review the expression-related code, so I'll neither approve nor reject. I looked at the refactor around `runBigQuery` etc and that all looks good to me.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala [hunk 1]
```diff
@@ -176,53 +293,15 @@ trait DataRepoBigQuerySupport {
...
-   * Google's doc on column naming (https://cloud.google.com/bigquery/docs/schemas) says:
-   *   - A column name must contain only letters (a-z, A-Z), numbers (0-9), or underscores (_), and it must start with a letter or underscore. The maximum column name length is 128 characters
-   *
-   * Given those naming requirements, this function allows ONLY: letters, numbers, underscores, hyphens.
-   * It does not impose any length restrictions.
-   *
-   * @param input the string to be validated
-   * @return the original string, if valid; throws an error if invalid.
-   */
-  def validateSql(input: String): String = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala:218]: moved to companion object

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala [hunk 2]
```diff
@@ -176,53 +293,15 @@ trait DataRepoBigQuerySupport {
    * @param entityQuery user-supplied query criteria
    * @return the object to pass to BQ to execute a query
    */
-  def queryConfigForQueryEntities(dataProject: String, viewName: String, entityType: String, entityQuery: EntityQuery): QueryJobConfiguration = {
+  def queryConfigForQueryEntities(dataProject: String, viewName: String, entityType: String, entityQuery: EntityQuery): QueryJobConfiguration.Builder = {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala:297]: this seemingly unrelated change (and similar that follow) allows us to set maximumBytesBilled centrally

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala [hunk 3]
```diff
@@ -76,25 +78,20 @@ class DataRepoEntityProvider(snapshotModel: SnapshotModel, requestArguments: Ent
...
-      .build
-
-    // get pet service account key for this user
-    samDAO.getPetServiceAccountKeyForUser(googleProject, requestArguments.userInfo.userEmail) map { petKey =>
-
-      // get a BQ service (i.e. dao) instance, and use it to execute the query against BQ
-      val queryResource: Resource[IO, TableResult] = for {
-        bqService <- bqServiceFactory.getServiceForPet(petKey)
-        queryResults <- Resource.liftF(bqService.query(queryConfig))
-      } yield queryResults
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:90]: this has been restructured to be in terms of `IO` and not a nested map under `samDAO.getPetServiceAccountKeyForUser`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala [hunk 4]
```diff
@@ -280,61 +336,16 @@ class DataRepoEntityProvider(snapshotModel: SnapshotModel, requestArguments: Ent
...
-        val viewName = snapshotModel.getName
-        // generate BQ SQL for this entity
-        val query = s"SELECT $baseTableAlias.$entityNameColumn, ${expressions.map(_.qualifiedColumnName).mkString(", ")} FROM `${dataProject}.${viewName}.${tableModel.getName}` $baseTableAlias"
-
-        (expressions, QueryJobConfiguration.newBuilder(query).build)
-
-      case _ => throw new RawlsException("relations not implemented yet")
-    }
-  }
-
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:326]: moved to `DataRepoBigQuerySupport`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoEvaluateToAttributeVisitor.scala [hunk 5]
```diff
@@ -18,19 +16,7 @@ class DataRepoEvaluateToAttributeVisitor(rootTableAlias: String) extends TerraEx
...
-    } else {
-      relations.last.getText
-    }
-
-    Seq(ParsedEntityLookupExpression(
-      relations.map(_.attributeName().getText),
-      ctx.attributeName().getText.toLowerCase,
-      ctx.getText,
-      tableAlias
-    ))
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/expressions/parser/antlr/DataRepoEvaluateToAttributeVisitor.scala:33]: figured out better aliasing so removed alias stuff in this file

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala [hunk 6]
```diff
@@ -1,18 +1,124 @@
...
+    * Google's doc on column naming (https://cloud.google.com/bigquery/docs/schemas) says:
+    *   - A column name must contain only letters (a-z, A-Z), numbers (0-9), or underscores (_), and it must start with a letter or underscore. The maximum column name length is 128 characters
+    *
+    * Given those naming requirements, this function allows ONLY: letters, numbers, underscores, hyphens.
+    * It does not impose any length restrictions.
+    *
+    * @param input the string to be validated
+    * @return the original string, if valid; throws an error if invalid.
+    */
+  def validateSql(input: String): String = {
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala:54]: I'm not sure where else we're dynamically generating BigQuery queries, but at some point it might make sense to have this in a shared lib.

Should we highlight this function explicitly with the AppSec team?  I am pretty sure they run scans for SQL Injection periodically, but I' wondering how different BigQuery is and whether they need/want to write specific scenarios to test how well we're sanitizing strings that get used in queries.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala [hunk 7]
```diff
@@ -76,25 +78,20 @@ class DataRepoEntityProvider(snapshotModel: SnapshotModel, requestArguments: Ent
     // they should be safe, but we should have layers of protection.
     val query = s"SELECT * FROM `${validateSql(dataProject)}.${validateSql(viewName)}.${validateSql(entityType)}` WHERE $pk = @pkvalue;"
     // generate query config, with named param for primary key
-    val queryConfig = QueryJobConfiguration.newBuilder(query)
+    val queryConfigBuilder = QueryJobConfiguration.newBuilder(query)
+      .setMaximumBytesBilled(config.bigQueryMaximumBytesBilled)
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala]: Can remove this now that this is set centrally in `runBigQuery`

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala [hunk 8]
```diff
@@ -164,36 +139,82 @@ class DataRepoEntityProvider(snapshotModel: SnapshotModel, requestArguments: Ent
...
+    def stringSize(s: String) = objectSize + s.length * 1
+
+    def sizeOfJsValueBytes(jsValue: JsValue): Int = {
+      jsValue match {
+        case JsArray(elements) => objectSize + elements.map(sizeOfJsValueBytes).sum
+        case JsBoolean(_) => booleanSize
+        case JsNull => 0
+        case JsNumber(_) => bigDecimalSize
+        case JsObject(fields) => fields.map { case (k, v) => stringSize(k) + sizeOfJsValueBytes(v) }.sum
+        case JsString(value) => objectSize + stringSize(value)
```
Comment (marctalbott) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:156]: is `objectSize` being double counted here and on lines 166 & 167 or does the `JsString` and other case classes have size of their own?

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala [hunk 9]
```diff
@@ -1,18 +1,124 @@
...
+  }
+  case class EntityColumn(table: EntityTable, column: String, isArray: Boolean) {
+    lazy val qualifiedName = s"${validateSql(table.alias)}.${validateSql(column)}"
+  }
+
+  /**
+    * Contains the data to construct a JOIN in SQL and identify the JOIN in the results
+    * @param from left side of join
+    * @param to right side of join
+    * @param relationshipPath list of relationship names that resulted in this join
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala]: I'm not clear what we mean by the description of this param

**Review by marctalbott — APPROVED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala (hunk 8 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:156]: JsString is an object so has its own size. We might be undercounting big decimals but that is a back of the envelop guesstimate anyway.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala [hunk 10]
```diff
@@ -234,4 +314,266 @@ trait DataRepoBigQuerySupport {
...
+    * @param selectAndFroms
+    * @return
+    */
+  private[datarepo] def generateExpressionSQL(selectAndFroms: Seq[SelectAndFrom]) = {
+    // the head of selectAndFroms is special, it is the starting point of the query
+    // all elements of the tail eventually join back to the head (like that snake thing)
+    val rootTableJoin = selectAndFroms.head
+
+    if (rootTableJoin.join.isDefined) {
+      throw new DataEntityException(s"expected rootTableJoin.relationship to be None")
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala]:
```suggestion
      throw new DataEntityException(s"expected rootTableJoin.join to be None")
```

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala [hunk 11]
```diff
@@ -234,4 +314,266 @@ trait DataRepoBigQuerySupport {
...
+      // getting 1 row per entity and still run 1 workflow per entity
+      val idColumn = EntityColumn(rootEntityTable, entityNameColumn, false)
+      Seq(SelectAndFrom(rootEntityTable, None, Seq(idColumn)))
+    } else {
+      figureOutQueryStructureForExpressions(snapshotModel, rootEntityTable, parsedExpressions, entityNameColumn)
+    }
+
+    val query: String = generateExpressionSQL(selectAndFroms)
+
+    (selectAndFroms, QueryJobConfiguration.newBuilder(query))
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala:339]: I'm not sure where to put it exactly, but somewhere around here I feel like we should have a log statement to tell us the list of expressions that we evaluated and the resulting BQ query that we generated.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala [hunk 12]
```diff
@@ -249,28 +273,64 @@ class DataRepoEntityProvider(snapshotModel: SnapshotModel, requestArguments: Ent
...
-        queryResults <- bqQueryJobConfigs.toList.traverse { // consider parTraverse if serial is slow but be careful not to use too many threads
-          case (expressions, bqJob) => bqService.query(bqJob).map(expressions -> _)
-        }
-        _ <- checkQuerySize(queryResults)
-      } yield {
-        transformQueryResultToExpressionAndResult(entityNameColumn, queryResults)
-      }
-    }
+  private def runBigQuery(bqQueryJobConfigBuilder: QueryJobConfiguration.Builder, petKey: String): IO[TableResult] = {
+    bqServiceFactory.getServiceForPet(petKey).use(_.query(
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:278]: When do we want to log queries and when do we not?  I think logging queries is going to be really helpful for debugging.  We probably want to log at the level of the caller of this method, so we can log the query with some context too, but logging here would be better than nothing.

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala [hunk 13]
```diff
@@ -249,28 +273,64 @@ class DataRepoEntityProvider(snapshotModel: SnapshotModel, requestArguments: Ent
...
+      bqQueryJobConfigBuilder
+        .setMaximumBytesBilled(config.bigQueryMaximumBytesBilled)
+        .build()))
   }
 
-  private def transformQueryResultToExpressionAndResult(entityNameColumn: String, queryResults: List[(Set[ParsedEntityLookupExpression], TableResult)]): List[ExpressionAndResult] = {
+  /**
+    * The tableResult should have a single row per root entity. Each row should have a column for each
```
Comment (gpolumbo-broad) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:285]: This first sentence makes a lot of sense now that I understand things more.  Can we use this to our advantage is estimating sizes of result sets?  For example, in the simplest case, if the root table has more rows than our threshold memory size, we can preemptively alert the user.  Smarter thing would be to do some rough calculation based on maybe the number of unique expressions and the smallest data type size, that sort of thing.  I guess it would boil down to asking whether it would be worthwhile to preemptively alert users that their queries may fail based on the fact that their root table is so large.

**Review by gpolumbo-broad — CHANGES_REQUESTED** [review]

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala (hunk 9 above)
Comment (gpolumbo-broad) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala]: this was an old comment that has now been resolved

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala (hunk 12 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:278]: workbench-libs logs all the queries

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala (hunk 13 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoEntityProvider.scala:285]: we already do this, see the function `checkSubmissionSize`

**Comment by dvoet** [comment]
jenkins retest

File: core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala (hunk 6 above)
Comment (dvoet) [reply core/src/main/scala/org/broadinstitute/dsde/rawls/entities/datarepo/DataRepoBigQuerySupport.scala:54]: this was written by App Services... I would put this on their plate

**Review by gpolumbo-broad — APPROVED** [review]
